File: leetcode/hash tables/1160. Find Words That Can Be Formed by Characters/golang/main.go
package main

import (
	"fmt"
)

func countCharacters(words []string, chars string) int {
	ans := 0

	charsCnt := make(map[rune]int)
	for _, r := range chars {
		charsCnt[r]++
	}

	for _, w := range words {
		wordCnt := make(map[rune]int)
		ok := true

		for _, r := range w {
			wordCnt[r]++
			if wordCnt[r] > charsCnt[r] {
				ok = false
				break
			}
		}

		if ok {
			ans += len([]rune(w))
		}
	}

	return ans
}

func main() {
	words := []string{"cat", "bt", "hat", "tree"}
	chars := "atach"
	fmt.Println(countCharacters(words, chars))
}

File: leetcode/hash tables/1160. Find Words That Can Be Formed by Characters/python/main.py
"""
------------------------------
------------------------------

Example 1:

Input: words = ["cat","bt","hat","tree"], chars = "atach"
Output: 6
Explanation: The strings that can be formed are "cat" and "hat" so the answer is 3 + 3 = 6.
Example 2:

Input: words = ["hello","world","leetcode"], chars = "welldonehoneyr"
Output: 10
Explanation: The strings that can be formed are "hello" and "world" so the answer is 5 + 5 = 10.

------------------------------
------------------------------
"""

from typing import Counter, List


class Solution:
    def countCharacters(self, words: List[str], chars: str) -> int:
        ans, charts_cnt = 0, Counter(chars)

        for w in words:
            ans += len(w) if charts_cnt >= Counter(w) else 0

        return ans


words = ["cat", "bt", "hat", "tree"]
chars = "atach"
print(Solution().countCharacters(words, chars))
